// o1js/dist/web/lib/mina/mina-instance.js
import { UInt64 } from '../int.js';
export { defaultNetworkConstants, activeInstance, setActiveInstance, ZkappStateLength, };
const defaultAccountCreationFee = 1000000000;
const defaultNetworkConstants = {
    genesisTimestamp: UInt64.from(0),
    slotTime: UInt64.from(3 * 60 * 1000),
    accountCreationFee: UInt64.from(defaultAccountCreationFee),
};
const ZkappStateLength = 8;
let activeInstance = {
    accountCreationFee: () => defaultNetworkConstants.accountCreationFee,
    getNetworkConstants: () => defaultNetworkConstants,
    currentSlot: noActiveInstance,
    hasAccount: noActiveInstance,
    getAccount: noActiveInstance,
    getNetworkState: noActiveInstance,
    sendTransaction: noActiveInstance,
    transaction: noActiveInstance,
    fetchEvents: noActiveInstance,
    fetchActions: noActiveInstance,
    getActions: noActiveInstance,
    proofsEnabled: true,
    getNetworkId: () => 'testnet',
};
/**
 * Set the currently used Mina instance.
 */
function setActiveInstance(m) {
    activeInstance = m;
}
function noActiveInstance() {
    throw Error('Must call Mina.setActiveInstance first');
}
